modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/SourceConfirmedTextQuery.java

```diff
@@ -385,6 +394,20 @@ private float computeFreq() throws IOException {
             }
             return frequency;
         }
+
+        private Matches matches() throws IOException {
+            MemoryIndex index = new MemoryIndex(true, false);
```

Comment:
I wonder if we can avoid the duplicated memory index here. 
When the field is indexed without offset the unified highlighter will use a memory index to reindex with offsets and positions. 
Maybe we can rewrite the `SourceConfirmedTextQuery` into its child query when we detect that the field has positions and offsets enabled?